File: iosApp/iosApp/Features/Wallets/WalletsViewModel.swift
import Foundation

/// Wallets state matching Android's WalletsState
struct WalletsState {
    var wallets: [Wallet] = []
    var isLoading: Bool = false
    var error: String? = nil
}

/// Wallets events matching Android's WalletsEvent
enum WalletsEvent {
    case loadWallets
    case refresh
    case clearError
}

/// WalletsViewModel matching Android's WalletsViewModel
class WalletsViewModel: BaseViewModel<WalletsState, WalletsEvent> {
    private let walletRepository: WalletRepositoryProtocol
    private let navigationCoordinator: NavigationCoordinator

    init(
        walletRepository: WalletRepositoryProtocol,
        navigationCoordinator: NavigationCoordinator
    ) {
        self.walletRepository = walletRepository
        self.navigationCoordinator = navigationCoordinator
        super.init(initialState: WalletsState())
        loadWallets()
    }

    override func onEvent(_ event: WalletsEvent) {
        switch event {
        case .loadWallets, .refresh:
            loadWallets()
        case .clearError:
            state.error = nil
        }
    }

    private func loadWallets() {
        Task { [weak self] in
            guard let self = self else { return }

            await MainActor.run {
                self.state.isLoading = true
                self.state.error = nil
            }

            do {
                let wallets = try await walletRepository.getWallets()
                await MainActor.run {
                    self.state.wallets = wallets
                    self.state.isLoading = false
                }
            } catch {
                await MainActor.run {
                    self.state.error = "Failed to load wallets: \(error.localizedDescription)"
                    self.state.isLoading = false
                }
            }
        }
    }
}

File: androidApp/src/main/java/org/idos/app/ui/screens/wallets/WalletsViewModel.kt
package org.idos.app.ui.screens.wallets

import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import org.idos.app.data.model.Wallet
import org.idos.app.data.repository.WalletRepository
import org.idos.app.ui.screens.base.BaseViewModel
import timber.log.Timber

sealed class WalletsEvent {
    object LoadWallets : WalletsEvent()
    object ClearError : WalletsEvent()
}

data class WalletsState(
    val wallets: List<Wallet> = emptyList(), val isLoading: Boolean = true, val error: String? = null
)

class WalletsViewModel(
    private val walletRepository: WalletRepository
) : BaseViewModel<WalletsState, WalletsEvent>() {

    override fun initialState(): WalletsState = WalletsState()

    override fun onEvent(event: WalletsEvent) {
        when (event) {
            is WalletsEvent.LoadWallets -> loadWallets()
            WalletsEvent.ClearError -> clearError()
        }
    }

    private fun loadWallets() {
        viewModelScope.launch {
            updateState { copy(isLoading = true, error = null) }
            try {
                walletRepository.getWallets().collectLatest { wallets ->
                    Timber.d("have wallets {$wallets}")
                    updateState {
                        copy(
                            wallets = wallets, isLoading = false
                        )
                    }
                }
            } catch (e: Exception) {
                Timber.d("no wallets, have exception")
                updateState {
                    copy(
                        isLoading = false, error = "Failed to load wallets: ${e.message}"
                    )
                }
            }
        }
    }

    private fun clearError() {
        updateState { copy(error = null) }
    }
}
